### ISSUE TITLE
Investigate cutadapt -j flag

### ISSUE THREAD

**Issue by mikej888** [riboviz/RiboViz#43]
[cutadapt](https://github.com/marcelm/cutadapt) 2.x has a `-j` flag which allows the number of processors available to be specified (or automatically deduced). The [documentation](https://cutadapt.readthedocs.io/en/stable/guide.html) states:

> Make also sure that you have pigz (parallel gzip) installed if you use multiple cores and write to a .gz output file.

Check use of `-j` flag on Ubuntu and CentOS VMs, existence of pigz library, and effect this flag has on times to run cutadapt.

**Comment by mikej888** [comment]
My current Ubuntu 18 and Centos 7.6 both have:

```console
$ python --version
Python 2.7.16 :: Anaconda, Inc.
$ cutadapt --version
1.18
```
```console
$ python --version
Python 3.7.3
$ cutadapt --version
2.3
```

[pigz](http://zlib.net/pigz/) is a parallel implementation of gzip for multi-processor, multi-core machines.

```console
$ pigz --version
pigz 2.4
```

pigz was already present on both Ubuntu and CentOS VMs. The RiboViz user documentation can request user's check for it and, if it is not present, to install it via:

* Ubuntu:

```console
sudo apt-get install pigz
```

* CentOS:

```console
sudo yum install pigz
```

A sample cutadapt invocation by RiboViz is:

```
cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o vignette/tmp/WT3AT_trim.fq \
    vignette/input/SRR1042864_s1mi.fastq.gz -j 4
```

where, from cutadapt 2.3 [guide](https://cutadapt.readthedocs.io/en/v2.3/guide.html):

* `--trim-n`: N-end trimming, remove flanking `N` bases from each end post-adapter removal.
* `-O`: Minimum overlap length.
* `-m`: minimum length of output reads, those shorter before or after adapter removal are discarded.
* `-a`: adapter sequence, which is searched for and removed from all reads.
* `-o`: output file.
* `-j`: multi-core, needs `pigz` (parallel GZIP) otherwise output compression is done in a single-thread. In RiboViz, this value is specified via the `vignette_config.yaml` `nprocesses` parameter.

Python 2 run with no `-j` flag:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.fq  vignette/input/SRR1042864_s1mi.fastq.gz
This is cutadapt 1.18 with Python 2.7.16
Command line parameters: --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.fq vignette/input/SRR1042864_s1mi.fastq.gz
Processing reads on 1 core in single-end mode ...
Finished in 17.30 s (13 us/read; 4.77 M reads/minute).
```

Python 3 run with no `-j` flag:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.fq  vignette/input/SRR1042864_s1mi.fastq.gz
This is cutadapt 2.3 with Python 3.7.3
Command line parameters: --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.py3.fq vignette/input/SRR1042864_s1mi.fastq.gz
Processing reads on 1 core in single-end mode ...
Finished in 13.55 s (10 us/read; 6.09 M reads/minute).
```

Note in both cases 1 core is used.

Specify 4 cores,`-j 4`:

* Python 2 run:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.4.fq  vignette/input/SRR1042864_s1mi.fastq.gz -j 4
ERROR: Running in parallel is not supported on Python 2
```

* Python 3 run:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.4.fq  vignette/input/SRR1042864_s1mi.fastq.gz -j 4
This is cutadapt 2.3 with Python 3.7.3
Command line parameters: --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.4.fq vignette/input/SRR1042864_s1mi.fastq.gz -j 4
Processing reads on 4 cores in single-end mode ...
Finished in 4.79 s (3 us/read; 17.21 M reads/minute).
```

Note processing times and rates compared to `13.55 s (10 us/read; 6.09 M reads/minute)` above when `-j` was not specified and only 1 core was used.

Request auto-detect cores,`-j 0`:

* Python 2 run:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.0.fq  vignette/input/SRR1042864_s1mi.fastq.gz -j 0
ERROR: Running in parallel is not supported on Python 2
```

* Python 3 run:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.4.fq  vignette/input/SRR1042864_s1mi.fastq.gz -j 0
This is cutadapt 2.3 with Python 3.7.3
Command line parameters: --trim-n -O 1 -m 5 -a CTGTAGGCACC -o WT3AT_trim.4.fq vignette/input/SRR1042864_s1mi.fastq.gz -j 0
Processing reads on 4 cores in single-end mode ...
Finished in 4.48 s (3 us/read; 18.40 M reads/minute).
```

It would be good if cutadapt could assume that if `-j 0` is provided and Python 2 is being used then just treat this as if `-j` was not provided. However, the Python 2 version of cutadapt above was 1.18. cutadapt's changes for [v2.0 (2019-03-06)](https://cutadapt.readthedocs.io/en/stable/changes.html#v2-0-2019-03-06) comments that:

> This release of Cutadapt requires at least Python 3.4 to run. Python 2.7 is no longer supported.

More generally, Python 2 comes to the end of its supported life in 2020 (see [PEP 373 Python 2.7 Release Schedule](https://legacy.python.org/dev/peps/pep-0373/)) and there will be no Python 2.8.

There's no point in requesting a change in cutadapt. To provide a workaround for any existing RiboViz users who are using Python 2 we can include in prep_riboviz.py a check for the Python version and, if it's Python 2, set `-j 1` else set `-j 0`. The `sys` library can provide the version:

```python
>>> import sys
>>> sys.version_info
(2, 5, 2, 'final', 0)
>>> sys.version_info.major
2
```
```python
>>> import sys
>>> sys.version_info
sys.version_info(major=2, minor=7, micro=16, releaselevel='final', serial=0)
>>> sys.version_info.major
3
```

**Comment by mikej888** [comment]
Created branch [configure-cutadapt-cores-43](/riboviz/riboviz/tree/configure-cutadapt-cores-43).

**Comment by FlicAnderson** [comment]
In case it's useful: running `pigz --version` on my reasonably fresh Ubuntu 18.04 LTS install showed I've got pigz 2.4 already installed.  

Since I didn't know about it until yesterday I can safely say I didn't intentionally install it, so it's either come with Ubuntu by default or was pulled in as a dependency on something else during riboviz bits-and-bobs installation.

**Comment by mikej888** [comment]
It's not on my "vanilla" Ubuntu 18 or CentOS 7 VMs so I'd agree it's pulled in by some RiboViz dependency's dependency!

**Comment by mikej888** [comment]
Implemented in 17a2a06 and merged into develop.
